{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#competition.m\n",
    "\n",
    "def competition(oldPop,A,B,C):\n",
    "    totalPop = sum(oldPop)\n",
    "    newPop = newPop = A*oldPop/(1+B+C*totalPop)\n",
    "\n",
    "    return newPop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# deeptill.m\n",
    "\n",
    "def deepTill(upperBank,lowerBank,tillFrequency,currentYear):\n",
    "    if tillFrequency(currentYear) == 1:\n",
    "        tmp = upperBank\n",
    "        upperBank = lowerBank\n",
    "        lowerBank = tmp\n",
    "\n",
    "    return upperBank, lowerBank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extinction.m\n",
    "\n",
    "import random\n",
    "from math import inf\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def extinction(oldPopulation,fieldSize):\n",
    "    if fieldSize == inf:\n",
    "        newPopulation = oldPopulation\n",
    "\n",
    "    elif fieldSize == 0:\n",
    "        raise \"old pop returned\"\n",
    "\n",
    "    else:\n",
    "        # %         rng('shuffle');\n",
    "        nGeno = len(oldPopulation);\n",
    "        rndExtinct = random.random(1,nGeno);\n",
    "        newPopulation = np.zeros(1,nGeno);\n",
    "\n",
    "        for j in range(nGeno):\n",
    "            if oldPopulation[j] == 0:\n",
    "                # %first check if jth genotype population is 0\n",
    "                newPopulation[j] = 0\n",
    "            elif oldPopulation[j] >= 1/fieldSize:\n",
    "                # %next check if jth genotype number in field >=1\n",
    "                newPopulation[j] = oldPopulation(j)\n",
    "            elif rndExtinct[j] > 0.5:\n",
    "                # %jth genotype survives extinction!\n",
    "                newPopulation[j] = 1/fieldSize\n",
    "            else:\n",
    "                # %otherwise, jth genotype dies...\n",
    "                newPopulation[j] = 0\n",
    "\n",
    "    return newPopulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#getresistancetime\n",
    "\n",
    "def getResistanceTime(time,res_frac,f,varargin):\n",
    "    # %time is a (Mx1) array. res_frac is an (Mx1) array. time and res_frac\n",
    "    # %must have the same number of rows, otherwise an error is returned. f\n",
    "    # %is the resistant fraction whose time-to-resistance will be found. f\n",
    "    # %may be a (Px1) array. Function returns the time-to-resistance, tR,\n",
    "    # %which is a (Px1) array. varargin is a Px1 array which may be used to\n",
    "    # %keep \"count\" of how many times resistance developed for each element\n",
    "    # %of f. If resistance is found for current value of f, a counter\n",
    "    # %variable is incremented by 1. This is then returned in counter. This\n",
    "    # %is useful for counting how many times resistance developed over many\n",
    "    # %function calls\n",
    "    \n",
    "    if len(time) != len(res_frac):\n",
    "        raise 'time and res_frac must have equal number of rows'\n",
    "        \n",
    "    else:\n",
    "        # %Counter variable\n",
    "        if varargin:\n",
    "            counter = varargin[1];\n",
    "        else:\n",
    "            counter = np.zeros(len(f),1)\n",
    "\n",
    "        \n",
    "        # %number of evaluation points, P\n",
    "        P = len(f)\n",
    "        \n",
    "        # %output variable, a (Px1) array\n",
    "        tR = np.zeros(P,1)\n",
    "        \n",
    "        # %Loop through f and find corresponding times.\n",
    "        for j in range(len(P)):\n",
    "\n",
    "            tmp = find(res_frac>f(j),1,'first')\n",
    "\n",
    "            if tmp == 1:\n",
    "                # %The initial resistance exceeded the threshold\n",
    "                tR[j] = 0;\n",
    "                \n",
    "                # %Increment counter\n",
    "                counter[j] = counter[j] + 1\n",
    "            elif tmp:\n",
    "                # %If the threshold is exceeded, interpolate to find the time\n",
    "\n",
    "                # %Temporary time and resistant fraction variables for\n",
    "                # %readability\n",
    "                t2 = time(tmp)\n",
    "                t1 = time(tmp-1)\n",
    "                f2 = res_frac(tmp)\n",
    "                f1 = res_frac(tmp-1)\n",
    "\n",
    "                # %Linear interpolation\n",
    "                tR[j] = (t2 - t1) / (f2 - f1) * (f[j] - f1) + t1\n",
    "                \n",
    "                # %Increment counter\n",
    "                counter[j] += 1\n",
    "            else:\n",
    "                # %The threshold was never exceeded\n",
    "                tR[j] = inf\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mating(population,s,key):\n",
    "    nLoci,nGeno = size(key);\n",
    "\n",
    "\n",
    "# %% Pure outcrossing\n",
    "    if sum(population) == 0:\n",
    "        # raise('Population is zero')\n",
    "        newFracOut = np.zeros(1,len(population))\n",
    "        \n",
    "    else:\n",
    "        nLoci = size(key,1);\n",
    "        nGeno = size(key,2);\n",
    "        newFracOut = ones(1,nGeno);\n",
    "\n",
    "        p = np.zeros(1,nLoci);\n",
    "\n",
    "        %determine current allele frequency\n",
    "        for j in range(nLoci):\n",
    "            for k in range(nGeno):\n",
    "                if key(j,k) == 2:\n",
    "                    p[j] = p[j] + 2*population[k]\n",
    "                elif key[j,k] == 1:\n",
    "                    p[j] = p[j] + population[k]\n",
    "\n",
    "        p = 0.5*p/sum(population)\n",
    "        q = 1 - p\n",
    "\n",
    "        # %determine new genotype fraction\n",
    "        for j in range(nLoci):\n",
    "            for k in range(nGeno):\n",
    "                if key(j,k) == 2:\n",
    "                    newFracOut[k] = newFracOut(k)*p(j)^2\n",
    "                elif key(j,k) == 1:\n",
    "                    newFracOut[k] = newFracOut(k)*2*p(j)*q(j)\n",
    "                else:\n",
    "                    newFracOut[k] = newFracOut(k)*q(j)^2\n",
    "\n",
    "    \n",
    "# %% Pure selfing\n",
    "selfingTable = dlmread(strcat('selfTable',num2str(nLoci),'.txt'));\n",
    "newFracSelf = population*selfingTable;\n",
    "\n",
    "# %% Total new fraction\n",
    "newFrac = s*newFracSelf + (1-s)*newFracOut;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function mutatedPopulation = mutation(population,muS2R,muR2S,targetLocus,key)\n",
    "\n",
    "mu = muR2S;\n",
    "nu = muS2R;\n",
    "\n",
    "# %define a new population array and initilize it to the current population\n",
    "mutatedPopulation = population;\n",
    "\n",
    "# %cycle through genotypes\n",
    "for j = 1:size(key,2)\n",
    "    # %Note: Since mutation is taken one locus at a time, each\n",
    "    # %genotype can only become 2 other genotypes and vice versa, i.e.\n",
    "    # %for locus 1, [1,2,1] -> [0,2,1] or [2,2,1] only. For each\n",
    "    # %genotype, we need to first locate the other two that it can\n",
    "    # %interact with\n",
    "    tmp = mod(j,3^(targetLocus-1))+1;\n",
    "\n",
    "    indSS = tmp;\n",
    "    indRS = indSS + 3^(targetLocus-1);\n",
    "    indRR = indRS + 3^(targetLocus-1);\n",
    "\n",
    "    if key(targetLocus,j) == 2\n",
    "        # %if the current genotype + locus has 2 resistant alleles\n",
    "        mutatedPopulation(indRR) = mutatedPopulation(indRR) - 2*mu*population(indRR) + mu^2*population(indRR);\n",
    "        mutatedPopulation(indRS) = mutatedPopulation(indRS) + 2*mu*population(indRR) - 2*mu^2*population(indRR);\n",
    "        mutatedPopulation(indSS) = mutatedPopulation(indSS) + mu^2*population(indRR);\n",
    "    elseif key(targetLocus,j) == 1\n",
    "        # %if the current genotype + locus has 1 resistant alleles\n",
    "        mutatedPopulation(indRR) = mutatedPopulation(indRR) + nu*population(indRS);\n",
    "        mutatedPopulation(indRS) = mutatedPopulation(indRS) - nu*population(indRS) - mu*population(indRS);\n",
    "        mutatedPopulation(indSS) = mutatedPopulation(indSS) + mu*population(indRS);    \n",
    "    else\n",
    "        # %if the current genotype + locus has 0 resistant alleles\n",
    "        mutatedPopulation(indRR) = mutatedPopulation(indRR) + nu^2*population(indSS);\n",
    "        mutatedPopulation(indRS) = mutatedPopulation(indRS) + 2*nu*population(indSS) - 2*nu^2*population(indSS);\n",
    "        mutatedPopulation(indSS) = mutatedPopulation(indSS) - 2*nu*population(indSS) + nu^2*population(indSS);  \n",
    "    end\n",
    "\n",
    "    population = mutatedPopulation;\n",
    "\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function plt_han = plotPopulation(LifeStage,key,varargin)\n",
    "# %Accepts a lifestage of the population structure and plots the population\n",
    "# %versus time. Required inputs are LifeStage (e.g. Population.germination)\n",
    "# %and the genotype key (e.g. Params.key). Optional argument is the plot\n",
    "# %type: \n",
    "# %'total' = total population density \n",
    "# %'totalR' = total population %possessing resistant allele (including\n",
    "# %heterozygotes) \n",
    "# %'totalRR' = total %population possessing 2 resistant alleles \n",
    "# %'fracR' = fraction of population containing resistant allele (including\n",
    "# %heterozygotes)\n",
    "# %'fracRR' = fraction of population containing 2 resistant alleles\n",
    "\n",
    "if isempty(varargin)\n",
    "    keyword = 'total';\n",
    "else\n",
    "    keyword = varargin{1};\n",
    "end\n",
    "\n",
    "[nYears,nGeno,nCohorts] = size(LifeStage);\n",
    "N = nYears*nCohorts;\n",
    "\n",
    "newPop = zeros(N,nGeno);\n",
    "\n",
    "k = 1;\n",
    "for y = 1:nYears\n",
    "    for c = 1:nCohorts\n",
    "        newPop(k,:) = LifeStage(y,:,c);\n",
    "        k = k + 1;\n",
    "    end\n",
    "end\n",
    "\n",
    "switch keyword\n",
    "    case 'total'\n",
    "        pltPop = sum(newPop,2);\n",
    "        time = 0:N-1;\n",
    "        time = time / nCohorts;\n",
    "        ylab = 'Population, ($pl/m^2$)';\n",
    "        ymax = Inf;\n",
    "    case 'totalR'\n",
    "        pltPop = sumResistant(newPop,key);\n",
    "        time = 0:N-1;\n",
    "        time = time / nCohorts;\n",
    "        ylab = 'Population, ($pl/m^2$)';\n",
    "        ymax = Inf;\n",
    "    case 'totalRR'\n",
    "        pltPop = sumResistant(newPop,key,'homozygous');\n",
    "        time = 0:N-1;\n",
    "        time = time / nCohorts;\n",
    "        ylab = 'Population, ($pl/m^2$)';\n",
    "        ymax = Inf;\n",
    "    case 'fracR'\n",
    "        tmp = sumResistant(newPop,key);\n",
    "        pltPop = tmp./sum(newPop,2);\n",
    "        time = 0:N-1;\n",
    "        time = time / nCohorts;\n",
    "        ylab = 'Fraction resistant';\n",
    "        ymax = 1;\n",
    "    case 'yearlyfracR'\n",
    "        newPop = newPop(nCohorts:nCohorts:end,:);\n",
    "        tmp = sumResistant(newPop,key);\n",
    "        pltPop = tmp./sum(newPop,2);\n",
    "        time = 0:nYears-1;\n",
    "        ylab = 'Fraction resistant';\n",
    "        ymax = 1;\n",
    "        \n",
    "end\n",
    "\n",
    "hold all\n",
    "plt_han = plot(time,pltPop,'.-');\n",
    "xlabel('Year','FontSize',20,'Interpreter','latex')\n",
    "ylabel(ylab,'FontSize',20,'Interpreter','latex')\n",
    "axis([0 max(time) 0 ymax])\n",
    "set(gca,'FontSize',15)\n",
    "hold off"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepareParams_2(Params):\n",
    "# %Takes the user defined parameter structure array created in readData and \n",
    "# %prepares three variables for the simulation: UpperSeedBank, LowerSeedBank \n",
    "# %and Params. UpperSeedBank and LowerSeedBank contain the initial population\n",
    "# %information.\n",
    "\n",
    "# %% General parameters\n",
    "\n",
    "    P.General = Params.General;\n",
    "    P.Init = Params.Init;\n",
    "    P.Cultivation = Params.Cultivation;\n",
    "    P.Germination = Params.Germination;\n",
    "    P.Hand = Params.Hand;\n",
    "    P.Mature = Params.Mature;\n",
    "    P.SeedProd = Params.SeedProd;\n",
    "    P.Mutation = Params.Mutation;\n",
    "    P.Predation = Params.Predation;\n",
    "    P.Winter = Params.Winter;\n",
    "    P.Herbicide1 = Params.Herbicide1;\n",
    "    P.Herbicide2 = Params.Herbicide2;\n",
    "    P.Herbicide3 = Params.Herbicide3;\n",
    "    P.Herbicide4 = Params.Herbicide4;\n",
    "\n",
    "# %Define a few of the general parameters locally for use in\n",
    "# %the function (not most computationally efficient, but improves\n",
    "# %readability)\n",
    "nLoci = Params.General.nLoci;\n",
    "\n",
    "# %Generate a key for relating each genotype to the number of resistant\n",
    "# %alleles it posseses (0,1, or 2) at each gene locus. (Subfunction contained\n",
    "# %in this file).\n",
    "[P.key,P.keyText] = generateKey(nLoci);\n",
    "\n",
    "\n",
    "# %% Initialize the upper seed bank\n",
    "\n",
    "# %Calculate the fraction of each genotype by calling the matingEquilibrium\n",
    "# %function (subfunction in this file).\n",
    "tmp = matingEquilibrium(P.Init.Upper.ResAlleleFreq,P.General.selfingCoeff,P.key);\n",
    "\n",
    "# %Populate UpperSeedBank \n",
    "P.General.UpperSeedBank = tmp*Params.Init.Upper.seedDensity;\n",
    "\n",
    "# %% Initialize the lower seed bank\n",
    "\n",
    "# %Calculate the fraction of each genotype by calling the matingEquilibrium\n",
    "# %function (subfunction in this file).\n",
    "tmp = matingEquilibrium(P.Init.Lower.ResAlleleFreq,P.General.selfingCoeff,P.key);\n",
    "\n",
    "# %Populate LowerSeedBank \n",
    "P.General.LowerSeedBank = tmp*Params.Init.Lower.seedDensity;\n",
    "\n",
    "\n",
    "\n",
    "# %% Subfunctions\n",
    "def generateKey(nLoci):\n",
    "# %create a \"key\" array with rows correpsonding to a single gene locus,\n",
    "# %columns corresponding to unique genotypes, and each element containing\n",
    "# %the number of resistant alleles that a genotype has at a given gene\n",
    "# %locus. Example: key = [  0,0,0,1,1,1,2,2,2 ;\n",
    "# %   0,1,2,0,1,2,0,1,2 ] corresponds to the 9 unique genotypes when 2\n",
    "# %   loci are considered. First column contains 0 resistant alleles at\n",
    "# %   loci 1 and loci 2.\n",
    "# %This function also creates a cell array which contains the same\n",
    "# %information in more readable text format. This way, the user can\n",
    "# %easily correlate columns in variable \"key\" to a genotype. Example:\n",
    "# %key_text = [S1S1S2S2, S1S1R2S2, S1S1R2R2, ...etc, just run the code to\n",
    "# %see the output.\n",
    "\n",
    "# %number of unique genotypes\n",
    "nGeno = 3^nLoci;\n",
    "\n",
    "# %initialize key array\n",
    "key = zeros(nLoci,nGeno);\n",
    "\n",
    "# %populate the key array using the function odometer\n",
    "tmp = zeros(nLoci,1);\n",
    "for k = 1:nGeno\n",
    "    key(:,k) = tmp;\n",
    "    tmp = odometer(tmp,2);\n",
    "\n",
    "key_text = 0;\n",
    "\n",
    "# %create the text version for easier viewing by user\n",
    "key_text = cell(1,nGeno)\n",
    "for k = 1:nGeno\n",
    "    tmp = ''\n",
    "    for j = 1:nLoci:\n",
    "        locus_string = num2str(j)\n",
    "        if key(j,k) == 0:\n",
    "            tmp = strcat(tmp,'S',locus_string,'S',locus_string)\n",
    "        elseif key(j,k) == 1\n",
    "            tmp = strcat(tmp,'R',locus_string,'S',locus_string)\n",
    "        else\n",
    "            tmp = strcat(tmp,'R',locus_string,'R',locus_string)\n",
    "\n",
    "\n",
    "\n",
    "    key_text{k} = tmp\n",
    "\n",
    "\n",
    "    \n",
    "def odometer(odo,base_num):\n",
    "# %% Odometer function\n",
    "# %This function counts using a user specified number base. A digit takes\n",
    "# %on integer values between 0 and base_num. This function increments the\n",
    "# %right-most digit, unless the right-most digit is at the maximum value,\n",
    "# %in which case it is set to the first element of base_num and the next\n",
    "# %element is incremented. For example: odo = [0,0,1] will become odo =\n",
    "# %[0,0,2]. A second function call on odo will produce odo = [0,1,0].\n",
    "# %This works like a standard car odometer, but with an arbitrary number\n",
    "# %system.\n",
    "\n",
    "    N = length(base_num);       %number of possible elements\n",
    "    Ndigits = length(odo);      %number of elements in odo\n",
    "\n",
    "    j = 1;\n",
    "    while j <= Ndigits\n",
    "        if odo(j) < base_num\n",
    "            odo(j) = odo(j)+1\n",
    "            break\n",
    "\n",
    "        odo(j) = 0\n",
    "        j = j + 1\n",
    "\n",
    "\n",
    "def  matingEquilibrium(p,selfCoeff,key):\n",
    "    q = 1-p;\n",
    "    nLoci = size(key,1);\n",
    "    nGeno = size(key,2);\n",
    "    populationFraction_outCross = ones(1,nGeno);\n",
    "    %     populationFraction_selfing = ones(1,nGeno);\n",
    "\n",
    "    for j = 1:nLoci\n",
    "        for k = 1:nGeno\n",
    "            if key(j,k) == 2\n",
    "                populationFraction_outCross(k) = populationFraction_outCross(k)*p(j)^2;\n",
    "            elif key(j,k) == 1\n",
    "                populationFraction_outCross(k) = populationFraction_outCross(k)*2*p(j)*q(j);\n",
    "            else:\n",
    "                populationFraction_outCross(k) = populationFraction_outCross(k)*q(j)^2;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function p = readData( )\n",
    "# %Reads data from text files created by Excel and stores them into a\n",
    "# %structure for the program to use.\n",
    "\n",
    "format long g;\n",
    "\n",
    "# %% Create struct to hold all the data.\n",
    "p = struct('General',{},'Init', {},'Germination', {}, 'Cultivation', {}, ...\n",
    "    'Hand', {}, 'Mature', {},'SeedProd', {},'Mutation', {}, 'Predation', {}, ...\n",
    "    'Winter', {}, 'Herbicide1', {}, 'Herbicide2', {}, 'Herbicide3', {}, 'Herbicide4', {});\n",
    "\n",
    "# %% Initiate static parameters.\n",
    "p(1).General.currentYear = 1;\n",
    "p(1).General.currentCohort = 1;\n",
    "\n",
    "# %% Get Windows user name to construct correct file paths\n",
    "username = getenv('USERNAME');\n",
    "\n",
    "# %% Read and store life cycle parameter data.\n",
    "path = strcat('C:\\Users\\', username, '\\Desktop\\Weed_Resistance_Data.txt');\n",
    "\n",
    "# %Read file to local variable\n",
    "params_in = dlmread(path);\n",
    "\n",
    "p(1).General.nSim = params_in(1,1);                # %number of simulations\n",
    "p(1).General.nYears = params_in(2,1);              # %number of years\n",
    "p(1).General.nCohorts = params_in(3,1);             #%number of cohorts\n",
    "p(1).General.nLoci = 4;                             #%number of gene loci\n",
    "p(1).General.nGeno = 3^p.General.nLoci;            # %number of genotypes\n",
    "p(1).General.fieldSize = params_in(5,1);           # %field size\n",
    "p(1).General.selfingCoeff = params_in(6,1);        # %selfing coefficient \n",
    "p(1).General.seedDelay = params_in(7,1);           # %seed delay\n",
    "p(1).General.femaleFrac = params_in(8,1);          # %fraction female\n",
    "p(1).Init.Upper.seedDensity = params_in(9,1);      # %initial upper bank density\n",
    "p(1).Init.Lower.seedDensity = params_in(10,1);      # %initial lower bank density\n",
    "\n",
    "p(1).Init.Upper.ResAlleleFreq = params_in( 11:2:17,1 );  # %initial upper resistant allele frequency\n",
    "p(1).Init.Lower.ResAlleleFreq = params_in( 12:2:18,1 );   #%initial lower resistant allele frequency\n",
    "\n",
    "p(1).General.f_RO = params_in(19,1);         # %latent time\n",
    "p(1).General.f_CR = params_in(20,1);    # %critical response\n",
    "p(1).General.f_TR = params_in(21,1);    # %complete resistance time\n",
    "\n",
    "row = 22;\n",
    "for i = 1:p.General.nYears\n",
    "    for j = 1:p.General.nCohorts\n",
    "        p(1).Germination.survivalFraction(i,j) = params_in(row, 1);\n",
    "        p(1).Cultivation.survivalFraction(i,j) = params_in(row, 2);\n",
    "        p(1).Hand.survivalFraction(i,j) = params_in(row, 3);\n",
    "        p(1).Mature.maxPlants(i,j) = params_in(row, 4);\n",
    "        p(1).Mature.cropParam(i,j) = params_in(row, 5);\n",
    "        p(1).Mature.weedCompetition(i,j) = params_in(row, 6);\n",
    "        p(1).SeedProd.maxYield(i,j) = params_in(row, 7);\n",
    "        p(1).SeedProd.cropParam(i,j) = params_in(row, 8);\n",
    "        p(1).SeedProd.weedCompetition(i,j) = params_in(row, 9);\n",
    "        p(1).Mutation.muR2S_locus1(i,j) = params_in(row, 10);\n",
    "        p(1).Mutation.muS2R_locus1(i,j) = params_in(row, 11);\n",
    "        p(1).Mutation.muR2S_locus2(i,j) = params_in(row, 12);\n",
    "        p(1).Mutation.muS2R_locus2(i,j) = params_in(row, 13);\n",
    "        p(1).Mutation.muR2S_locus3(i,j) = params_in(row, 14);\n",
    "        p(1).Mutation.muS2R_locus3(i,j) = params_in(row, 15);\n",
    "        p(1).Mutation.muR2S_locus4(i,j) = params_in(row, 16);\n",
    "        p(1).Mutation.muS2R_locus4(i,j) = params_in(row, 17);\n",
    "        p(1).Predation.survivalFraction(i,j) = params_in(row, 18);\n",
    "        p(1).Winter.uppersurvivalFraction(i,j) = params_in(row, 19);\n",
    "        p(1).Winter.lowersurvivalFraction(i,j) = params_in(row, 20);\n",
    "        row = row + 1;\n",
    "\n",
    "\n",
    "\n",
    "# %% Read and store tilling data.\n",
    "path = strcat('C:\\Users\\', username, '\\Desktop\\Tilling_Data.txt');\n",
    "p(1).General.TillingFreq = dlmread(path);\n",
    "\n",
    "# %% Read and store herbicide data.\n",
    "path = strcat('C:\\Users\\', username, '\\Desktop\\Herbicide_Data.txt');\n",
    "data = dlmread(path);\n",
    "\n",
    "data_rows = size(data, 1);\n",
    "row = 1;\n",
    "for i = {'Herbicide1', 'Herbicide2', 'Herbicide3', 'Herbicide4'}\n",
    "    if row > data_rows\n",
    "        break\n",
    "    end\n",
    "    p(1).(i{1}).targetLocus = data(row, 1);\n",
    "    for k = 1:p.General.nYears\n",
    "        for m = 1:p.General.nCohorts\n",
    "            p(1).(i{1}).efficacySS(k, m) = data(row, 2);\n",
    "            p(1).(i{1}).efficacyRS(k, m) = data(row,3);\n",
    "            p(1).(i{1}).efficacyRR(k, m) = data(row,4);\n",
    "            row = row + 1;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def [int_risk,res_risk,res_data,RoverT] = resistanceRisk(nSim)\n",
    "\n",
    "#Runs the Main weed resistance code nSim times, determining the resistance\n",
    "#risk. The time to resistance is found for every value specified in p_eval,\n",
    "#which must be numbers between 0 and 1 (time is linearly interpolated\n",
    "#between years). The def returns the population for each of nSim\n",
    "#simulations in res_frac, an array of resistance propability co\n",
    "\n",
    "\n",
    "## Run simulation nSim times\n",
    "\n",
    "#Life stage to consider (seed bank)\n",
    "lifestage = 'seedBank' \n",
    "\n",
    "for n = 1:nSim\n",
    "    \n",
    "    #Call Main resistance def.\n",
    "    #(Params stays the same for all sims, so only store it once)\n",
    "    if n == 1\n",
    "        [Population,Params] = Main_2() \n",
    "        RoverT = zeros(Params.General.nYears,nSim) \n",
    "        res_data = zeros(3,nSim) \n",
    "        res_risk = zeros(Params.General.nYears,1) \n",
    "        years = 0:Params.General.nYears-1 \n",
    "        \n",
    "        \n",
    "    else\n",
    "        [Population,~] = Main_2(Params) \n",
    "        \n",
    "     \n",
    "    \n",
    "    #Count the resistants in each year, and the total population in each\n",
    "    #year.\n",
    "    \n",
    "    #resistant population\n",
    "    res_pop = sumResistant(Population.(lifestage)(:,:, ),Params.key) \n",
    "    \n",
    "    #total population\n",
    "    tot_pop = sum(Population.(lifestage)(:,:, ),2) \n",
    "    \n",
    "    #resistant fraction stored in res_frac\n",
    "    RoverT(:,n) = res_pop./tot_pop \n",
    "    \n",
    "    #Store resistance metrics: t_latent, t_crit, and t_total\n",
    "    res_data(:,n) = Population.Res \n",
    "    \n",
    "    #Find the resistance risk as a def of time\n",
    "    ind = find( RoverT(:,n) >= Params.General.criticalResTime ) \n",
    "    res_risk(ind) = res_risk(ind) + 1 \n",
    "    \n",
    "    \n",
    "    \n",
    " \n",
    "\n",
    "#Convert res_risk to a fraction\n",
    "res_risk = res_risk / nSim \n",
    "\n",
    "#Find integrated res_risk\n",
    "int_risk = trapz(years,res_risk)/length(years) \n",
    "\n",
    "#Plot the results\n",
    "scrz = get(0,'ScreenSize') \n",
    "figure('Position',[scrz(1)+100,scrz(2)+100,scrz(3)*0.6,scrz(4)*0.6])\n",
    "\n",
    "#All simulations\n",
    "ax1 = axes('Position',[0.1,0.1,0.35,0.85]) \n",
    "hold all\n",
    "plot(years,RoverT) \n",
    "axis([0,years( ),0,1])\n",
    "\n",
    "#Resistance emergence\n",
    "ax1 = axes('Position',[0.6,0.1,0.35,0.85]) \n",
    "t = linspace(years(1),years( ),1000) \n",
    "plot(t,pchip(years,res_risk*100,t),'k','LineWidth',2) \n",
    "axis([0,years( ),0,100])\n",
    "\n",
    " \n",
    "\n",
    "## Subdefs\n",
    "def Rpop = sumResistant(pop,key)\n",
    "\n",
    "    #Determine number of gene loci and genotypes from key variable\n",
    "    [nLoci,nGeno] = size(key) \n",
    "    \n",
    "    #Determine number of years and cohorts from population variable\n",
    "    [nYears,~,nCohorts] = size(pop) \n",
    "    \n",
    "    #Initialize output array to 0\n",
    "    Rpop = zeros(nYears,nCohorts) \n",
    "    \n",
    "    #Sum all with resistant allele\n",
    "    for G = 1:nGeno\n",
    "        for L = 1:nLoci\n",
    "            if key(L,G) == 2 || key(L,G) == 1\n",
    "                Rpop(:,:) = Rpop(:,:) + squeeze(pop(:,G,:)) \n",
    "                break\n",
    "             \n",
    "         \n",
    "     \n",
    "    \n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function [ resdata ] = resMetrics( resdata, currentYear, seedPop, ...\n",
    "                                        key, onset, critical, complete )\n",
    "                                    \n",
    "%resMetrics updates weed resistance data using population data from the \n",
    "%current year of the simulation. It takes as parameters the resistance\n",
    "%metrics array, the current year, the mature population for the current\n",
    "%year, the key, and the user-defined threshold values (percentages at which\n",
    "%weed resistance is considered to have developed, reached a critical value,\n",
    "%and affected the whole population).\n",
    "\n",
    "R = sumResistant(seedPop(currentYear, :, end),key); %resistant population\n",
    "T = sum(seedPop(currentYear, :, end),2);    %total population (at end of year)\n",
    "\n",
    "%Checks if: 1) the fraction of the populattion that is resistant has exceeded \n",
    "%the onset value and 2) the time of resistance onset has not been set. If\n",
    "%both conditions are met, the time of resistance onset is set to the\n",
    "%current year.\n",
    "if R(end)/T(end) >= onset && resdata(1, 1) == 0\n",
    "    resdata(1, 1) = currentYear;\n",
    "end\n",
    "\n",
    "%Checks if: 1) the fraction of the populattion that is resistant has exceeded \n",
    "%the critical value and 2) the time of critical resistance has not been set. \n",
    "%If both conditions are met, the time of critical resistance is set to the\n",
    "%current year.\n",
    "if R(end)/T(end) >= critical && resdata(1, 2) == 0\n",
    "    resdata(1, 2) = currentYear;\n",
    "end\n",
    "\n",
    "%Checks if: 1) the majority of the population is resistant (according to \n",
    "%the value specified by the user) and 2) the time of complete resistance has\n",
    "%not been set. If both conditions are met, the time of complete resistance \n",
    "%is set to the current year.\n",
    "if R(end)/T(end) >= complete && resdata(1, 3) == 0\n",
    "    resdata(1, 3) = currentYear;\n",
    "end\n",
    "\n",
    "end\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function [metrics,data_output] = runSimulation(varargin)\n",
    "%% Initialization\n",
    "\n",
    "    %Create a program status bar \n",
    "    h = waitbar(0, 'Please wait', 'Name', 'Plot progress', ...\n",
    "    'CreateCancelBtn', 'setappdata(gcbf, ''canceling'', 1)');\n",
    "    setappdata(h, 'canceling', 0)\n",
    "\n",
    "\n",
    "    %Run Main function with no arguments once to get the params variable\n",
    "    %populated\n",
    "    if ~isempty(varargin)\n",
    "        Params = varargin{1};\n",
    "    else\n",
    "        tmp = readData();\n",
    "        Params = prepareParams_2(tmp);\n",
    "    end\n",
    "    \n",
    "    %Number of simulations to run\n",
    "    if isinf(Params.General.fieldSize)\n",
    "        nSim = 1;\n",
    "    else\n",
    "        nSim = Params.General.nSim;\n",
    "    end\n",
    "    \n",
    "    nYears = Params.General.nYears;\n",
    "    nCohorts = Params.General.nCohorts;\n",
    "    \n",
    "    %Frequency values to find the corresponding resistance time for.\n",
    "    %Includes three user defined values, as well as evenly spaced\n",
    "    %values from 0.05 to 0.95 fraction resistant.\n",
    "    f_RO = Params.General.f_RO;\n",
    "    f_CR = Params.General.f_CR;\n",
    "    f_TR = Params.General.f_TR;\n",
    "    f = 0.05:0.05:0.95;\n",
    "    f = [f_RO;f_CR;f_TR;0.01;f';0.99];\n",
    "    f_len = length(f);\n",
    "    \n",
    "    counter = zeros(f_len,1);\n",
    "    \n",
    "    %Variables to store results\n",
    "    %variable to locally store the seed bank population\n",
    "    data_store = zeros(nCohorts*nYears,nSim+1);\n",
    "    \n",
    "    %variable for sending to the excel file (includes upper and lower seed\n",
    "    %banks). NOTE: I retroactively added this variable. It probably would\n",
    "    %be better code to have data_store and data_output be combined into a\n",
    "    %single array.\n",
    "    data_output = struct('upperbank_freq',zeros(nCohorts*nYears,nSim+1),...\n",
    "                         'lowerbank_freq',zeros(nYears,nSim+1));\n",
    "    \n",
    "    %structure array for hold population metrics\n",
    "    metrics = struct('IRF',zeros(nSim,1),'IRF_stats',0,...\n",
    "                                         'probs',zeros(f_len,1),...\n",
    "                                         't2R',Inf(f_len,nSim),...\n",
    "                                         't2R_stats',Inf(f_len,5));\n",
    "    \n",
    "    %Make first column contain time values\n",
    "    time_step = 1/nCohorts;\n",
    "    data_store(:,1) = 0:time_step:nYears-time_step;\n",
    "    data_output.upperbank_freq(:,1) = 0:time_step:nYears-time_step;\n",
    "    data_output.lowerbank_freq(:,1) = 0:nYears-1;\n",
    "    \n",
    "    %% Perform nSim simulations, storing the results \n",
    "    for n = 1:nSim\n",
    "        \n",
    "        %Call main function\n",
    "        [ Population,~ ] = Main(Params);\n",
    "       \n",
    "        %% All data for direct export\n",
    "        %Sum over all genotypes to find the total resistant and total seed\n",
    "        %population for each cohort/year\n",
    "        R = sumResistant( Population.seedBank,Params.key );\n",
    "        T = squeeze( sum(Population.seedBank,2) );\n",
    "        R_lower = sumResistant( Population.lowerBank,Params.key );\n",
    "        T_lower = squeeze( sum(Population.lowerBank,2) );\n",
    "        \n",
    "        %Calculate fraction resistant for each year/cohort\n",
    "        f_res = R./T;\n",
    "        f_res_lower = (R_lower ./ T_lower);\n",
    "        \n",
    "        %Zero divided by zero errors produce NaN for f_res. Replace NaN by\n",
    "        %0, since NaN errors come when populations get small.\n",
    "        f_res( isnan(f_res) ) = 0;\n",
    "        f_res_lower( isnan(f_res_lower) ) = 0;\n",
    "                \n",
    "        %Data is in (nYears x nCohorts) arrays currently. Rearrange so that it\n",
    "        %is a linear array with (1 x nYears x nCohorts) elements of increasing\n",
    "        %time.\n",
    "        for c = 1:nCohorts\n",
    "            data_store(c:nCohorts:end,n+1) = f_res(:,c);\n",
    "            data_output.upperbank_freq(c:nCohorts:end,n+1) = f_res(:,c);            \n",
    "        end\n",
    "        \n",
    "        data_output.lowerbank_freq(:,n+1) = f_res_lower;  \n",
    "        \n",
    "%         plot(data_store(:,1),data_store(:,n+1),'.-')\n",
    "        \n",
    "        %Use the trapezoid rule to \"integrate\" the resistance fraction\n",
    "        %(built in data integration function \"trapz\")\n",
    "        metrics.IRF(n) = trapz( data_store(:,1),data_store(:,n+1) ) / ( nCohorts*nYears*time_step );\n",
    "        \n",
    "        %% Determine if user defined thresholds were exceeded and at what\n",
    "        %time\n",
    "        [metrics.t2R(:,n),counter] = getResistanceTime(data_store(:,1),data_store(:,n+1),f,counter);\n",
    "\n",
    "        %Update wait bar\n",
    "        waitbar(n / nSim)\n",
    "    end\n",
    "    \n",
    "    %Delete wait bar upon completion of loop\n",
    "    delete(h)\n",
    "    \n",
    "    %Convert resistance counts to a probability\n",
    "    metrics.probs = counter / nSim;\n",
    "        \n",
    "    %% Compute statistics\n",
    "    %compute min, 10th,50th,90th percentile, and max time to resistance for\n",
    "    %each resistance level\n",
    "    metrics.IRF_stats = mean(metrics.IRF);\n",
    "    \n",
    "        for k = 1:length(f)\n",
    "            %find only simulations where resistance developed\n",
    "            ind = find(~isinf(metrics.t2R(k,:)));\n",
    "            tmp_data = metrics.t2R(k,ind);\n",
    "            \n",
    "            if ~isempty( tmp_data )\n",
    "                %min\n",
    "                metrics.t2R_stats(k,1) = min( tmp_data );\n",
    "\n",
    "                %10th percentile\n",
    "                metrics.t2R_stats(k,2) = prctile( tmp_data,10 );\n",
    "\n",
    "                %50th percentile\n",
    "                metrics.t2R_stats(k,3) = prctile( tmp_data,50 );\n",
    "\n",
    "                %90th percentile\n",
    "                metrics.t2R_stats(k,4) = prctile( tmp_data,90 );\n",
    "\n",
    "                %max\n",
    "                metrics.t2R_stats(k,5) = max( tmp_data );\n",
    "            end\n",
    "\n",
    "        end\n",
    "    \n",
    "    \n",
    "    \n",
    "    %% Write data to Excel file\n",
    "    \n",
    "    %Get reference to open Excel sheet (still open from parameter entry) \n",
    "    h = actxGetRunningServer('Excel.Application');\n",
    "    myBook = h.Workbooks.Item('WeedResistance_ParameterEntryInterface.xlsm');\n",
    "    \n",
    "    %Write metrics to \"Results\" tab\n",
    "    mySheet = myBook.Sheets.Item('Results');\n",
    "    \n",
    "    %Clear old data (Excel is already doing this, so it is commented out\n",
    "    %here)\n",
    "%     mySheet.Range('F7:F13').Value = '';\n",
    "%     mySheet.Range('C17:D1017').Value = '';\n",
    "%     mySheet.Range('H8:M1017').Value = '';\n",
    "%     \n",
    "%     %Number of simulations\n",
    "%     mySheet.Range('F7').Value = nSim;\n",
    "    \n",
    "    %Probability of exceeding resistances thresholds\n",
    "    mySheet.Range('F8').Value = metrics.probs(1);\n",
    "    mySheet.Range('F9').Value = metrics.probs(2);\n",
    "    mySheet.Range('F10').Value = metrics.probs(3);\n",
    "    \n",
    "    %Median times to threshold\n",
    "    mySheet.Range('F11').Value = metrics.t2R_stats(1,3);\n",
    "    mySheet.Range('F12').Value = metrics.t2R_stats(2,3);\n",
    "    mySheet.Range('F13').Value = metrics.t2R_stats(3,3);\n",
    "    \n",
    "    %Average integrated risk\n",
    "    mySheet.Range('F14').Value = mean(metrics.IRF);\n",
    "    \n",
    "    %Individual simulation integrated risk\n",
    "    n = 1:nSim;\n",
    "    mySheet.Range( strcat('I18:I',num2str(17+nSim)) ).Value = n';\n",
    "    mySheet.Range( strcat('J18:J',num2str(17+nSim)) ).Value = metrics.IRF;\n",
    "    \n",
    "    %Time-to-resistance statistics\n",
    "    mySheet.Range( strcat('B18:B',num2str(14+length(f))) ).Value = f(4:end);\n",
    "    mySheet.Range( strcat('C18:G',num2str(14+length(f))) ).Value = metrics.t2R_stats(4:end,1:5);\n",
    "    \n",
    "    %Write raw data (upper and lower seed bank resistant fraction) to text\n",
    "    %file\n",
    "    dlmwrite('UpperBank.csv',data_output.upperbank_freq,',');\n",
    "    dlmwrite('LowerBank.csv',data_output.lowerbank_freq,',');\n",
    "    \n",
    "end\n",
    "\n",
    "function [tR,counter] = getResistanceTime(time,res_frac,f,varargin)\n",
    "    %time is a (Mx1) array. res_frac is an (Mx1) array. time and res_frac\n",
    "    %must have the same number of rows, otherwise an error is returned. f\n",
    "    %is the resistant fraction whose time-to-resistance will be found. f\n",
    "    %may be a (Px1) array. Function returns the time-to-resistance, tR,\n",
    "    %which is a (Px1) array. varargin is a Px1 array which may be used to\n",
    "    %keep \"count\" of how many times resistance developed for each element\n",
    "    %of f. If resistance is found for current value of f, a counter\n",
    "    %variable is incremented by 1. This is then returned in counter. This\n",
    "    %is useful for counting how many times resistance developed over many\n",
    "    %function calls\n",
    "    \n",
    "    if length(time) ~= length(res_frac)\n",
    "        error('time and res_frac must have equal number of rows');\n",
    "        \n",
    "    else\n",
    "        %Counter variable\n",
    "        if ~isempty(varargin)\n",
    "            counter = varargin{1};\n",
    "        else\n",
    "            counter = zeros(length(f),1);\n",
    "        end\n",
    "        \n",
    "        %number of evaluation points, P\n",
    "        P = length(f);\n",
    "        \n",
    "        %output variable, a (Px1) array\n",
    "        tR = zeros(P,1);\n",
    "        \n",
    "        %Loop through f and find corresponding times.\n",
    "        for j = 1:P\n",
    "            tmp = find(res_frac>f(j),1,'first');\n",
    "\n",
    "            if tmp == 1\n",
    "                %The initial resistance exceeded the threshold\n",
    "                tR(j) = 0;\n",
    "                \n",
    "                %Increment counter\n",
    "                counter(j) = counter(j) + 1;\n",
    "            elseif ~isempty(tmp)\n",
    "                %If the threshold is exceeded, interpolate to find the time\n",
    "\n",
    "                %Temporary time and resistant fraction variables for\n",
    "                %readability\n",
    "                t2 = time(tmp);\n",
    "                t1 = time(tmp-1);\n",
    "                f2 = res_frac(tmp);\n",
    "                f1 = res_frac(tmp-1);\n",
    "\n",
    "                %Linear interpolation\n",
    "                tR(j) = (t2 - t1) / (f2 - f1) * (f(j) - f1) + t1;\n",
    "                \n",
    "                %Increment counter\n",
    "                counter(j) = counter(j) + 1;\n",
    "            else\n",
    "                %The threshold was never exceeded\n",
    "                tR(j) = Inf;\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function runSimulation(varargin)\n",
    "# % Function simulates the weed life-cycle for specified number of\n",
    "# % simulations. Function reads parameter input Excel sheet in the same\n",
    "# % directory. The function writes simulation results to the same\n",
    "# % spreadsheet, as well as writing csv files containing population data. The\n",
    "# % user may pass the Params structure from a previous simulation, if\n",
    "# % desired.\n",
    "\n",
    "# %% Initialization & preliminary calculations\n",
    "\n",
    "    # %Create a program status bar to alert user how far along the simulation\n",
    "    # %is\n",
    "    h = waitbar(0, 'Please wait', 'Name', 'Plot progress', ...\n",
    "    'CreateCancelBtn', 'setappdata(gcbf, ''canceling'', 1)');\n",
    "    setappdata(h, 'canceling', 0)\n",
    "\n",
    "    # %Run Main function with no arguments once to get the Params variable\n",
    "    # %populated, unless user has included Params as an argument of the\n",
    "    # %runSimulation function (in varargin).\n",
    "    if ~isempty(varargin)\n",
    "        Params = varargin{1};\n",
    "    else\n",
    "        tmp = readData();\n",
    "        Params = prepareParams_2(tmp);\n",
    "    end\n",
    "    \n",
    "    # %Store the number of simulations desired in local variable. If the\n",
    "    # %fieldsize is infinite, simulation is deterministic, so nSim is set to\n",
    "    # %1.\n",
    "    if isinf(Params.General.fieldSize)\n",
    "        nSim = 1;\n",
    "    else\n",
    "        nSim = Params.General.nSim;\n",
    "    end\n",
    "    \n",
    "    %Store the number of years and cohorts in local variables\n",
    "    nYears = Params.General.nYears;\n",
    "    nCohorts = Params.General.nCohorts;\n",
    "    \n",
    "    %Frequency values to find the corresponding resistance time for.\n",
    "    %Includes three user defined values.\n",
    "    f_RO = Params.General.f_RO;     %resistance onset frequency\n",
    "    f_CR = Params.General.f_CR;     %critical resistance frequency\n",
    "    f_TR = Params.General.f_TR;     %total resistance frequency\n",
    "    f = [f_RO;f_CR;f_TR];           %array containing 3 values\n",
    "    f_len = length(f);              %length of array\n",
    "    \n",
    "    %Counter variable for keeping track of the number of times resistance\n",
    "    %thresholds are exceeded. \n",
    "    counter = zeros(f_len,1);\n",
    "    \n",
    "    %Variables to store results\n",
    "    \n",
    "    %Variable to locally store the seed bank population\n",
    "    data_store = zeros(nCohorts*nYears,nSim+1);\n",
    "    \n",
    "    # %Variables for sending to the csv files (includes upper and lower seed\n",
    "    # %banks). Both resistance frequency and total density are stored and\n",
    "    # %written to files. NOTE: I retroactively added this variable. It\n",
    "    # %probably would be better code to have data_store and freq_output be\n",
    "    # %combined into a single array.\n",
    "    freq_output = struct('upperbank_freq',zeros(nCohorts*nYears,nSim+1),...\n",
    "                         'lowerbank_freq',zeros(nYears,nSim+1));\n",
    "    dens_output = struct('upperbank_freq',zeros(nCohorts*nYears,nSim+1),...\n",
    "                         'lowerbank_freq',zeros(nYears,nSim+1));\n",
    "    \n",
    "    # %Make first columns contain time values\n",
    "    time_step = 1/nCohorts;\n",
    "    data_store(:,1) = 0:time_step:nYears-time_step;\n",
    "    freq_output.upperbank_freq(:,1) = 0:time_step:nYears-time_step;\n",
    "    freq_output.lowerbank_freq(:,1) = 0:nYears-1;\n",
    "    dens_output.upperbank_freq(:,1) = 0:time_step:nYears-time_step;\n",
    "    dens_output.lowerbank_freq(:,1) = 0:nYears-1;\n",
    "    \n",
    "    # %Structure array for holding population metrics\n",
    "    metrics = struct('IRF',zeros(nSim,1),'IRF_stats',0,...\n",
    "                                         'probs',zeros(f_len,1),...\n",
    "                                         't2R',Inf(f_len,nSim),...\n",
    "                                         't2R_stats',Inf(f_len,5),...\n",
    "                                         'yearly_freq',{cell(nYears*nCohorts,6)} );\n",
    "    \n",
    "\n",
    "    \n",
    "%% Perform nSim simulations, storing the results \n",
    "    for n = 1:nSim\n",
    "        \n",
    "        %Call main function and store Population structure array\n",
    "        [ Population,~ ] = Main(Params);\n",
    "       \n",
    "    %% Compute metrics\n",
    "        %Sum over all genotypes to find the total resistant and total seed\n",
    "        %population for each cohort/year in the upper and lower banks\n",
    "        R = sumResistant( Population.seedBank,Params.key );\n",
    "        T = squeeze( sum(Population.seedBank,2) );\n",
    "        R_lower = sumResistant( Population.lowerBank,Params.key );\n",
    "        T_lower = squeeze( sum(Population.lowerBank,2) );\n",
    "        \n",
    "        %Calculate fraction resistant for each year/cohort\n",
    "        f_res = R./T;\n",
    "        f_res_lower = (R_lower ./ T_lower);\n",
    "        \n",
    "        %0/0 division errors produce NaN for f_res. Replace NaN by\n",
    "        %0, since NaN errors come when populations approach 0.\n",
    "        f_res( isnan(f_res) ) = 0;\n",
    "        f_res_lower( isnan(f_res_lower) ) = 0;\n",
    "                \n",
    "        %Data is in (nYears x nCohorts) arrays currently. Rearrange so that it\n",
    "        %is a linear array with (1 x nYears x nCohorts) elements of increasing\n",
    "        %time.\n",
    "        for c = 1:nCohorts\n",
    "            data_store(c:nCohorts:end,n+1) = f_res(:,c);\n",
    "            freq_output.upperbank_freq(c:nCohorts:end,n+1) = f_res(:,c);    \n",
    "            dens_output.upperbank_freq(c:nCohorts:end,n+1) = T(:,c);  \n",
    "        end\n",
    "        \n",
    "        %Store lower bank data\n",
    "        freq_output.lowerbank_freq(:,n+1) = f_res_lower;  \n",
    "        dens_output.lowerbank_freq(:,n+1) = T_lower;  \n",
    "                \n",
    "        %Use the trapezoid rule to \"integrate\" the resistance fraction\n",
    "        %(built in data integration function \"trapz\")\n",
    "        metrics.IRF(n) = trapz( data_store(:,1),data_store(:,n+1) ) / ( nCohorts*(nYears-1)*time_step );\n",
    "        \n",
    "        %% Determine if user defined thresholds were exceeded and at what\n",
    "        %time\n",
    "        [metrics.t2R(:,n),counter] = getResistanceTime(data_store(:,1),data_store(:,n+1),f,counter);\n",
    "\n",
    "        %Update wait bar\n",
    "        waitbar(n / nSim)\n",
    "    end\n",
    "\n",
    "%% Post-processing of simulation results\n",
    "    %Delete wait bar upon completion of loop\n",
    "    delete(h)\n",
    "    \n",
    "    %Convert resistance counts to a probability\n",
    "    metrics.probs = counter / nSim;\n",
    "        \n",
    "    %% Compute statistics\n",
    "    \n",
    "    %Find the mean of the integrated resistance fraction from each\n",
    "    %simulation\n",
    "    metrics.IRF_stats = mean(metrics.IRF);\n",
    "    \n",
    "    %Compute min, 10th,50th,90th percentile, and max times required to\n",
    "    %achieve the user specified resistance thresholds    \n",
    "        for k = 1:length(f)\n",
    "            %find only simulations where resistance developed\n",
    "            ind = find(~isinf(metrics.t2R(k,:)));\n",
    "            tmp_data = metrics.t2R(k,ind);\n",
    "            \n",
    "            if ~isempty( tmp_data )\n",
    "                %min\n",
    "                metrics.t2R_stats(k,1) = min( tmp_data );\n",
    "\n",
    "                %10th percentile\n",
    "                metrics.t2R_stats(k,2) = prctile( tmp_data,10 );\n",
    "\n",
    "                %50th percentile\n",
    "                metrics.t2R_stats(k,3) = prctile( tmp_data,50 );\n",
    "\n",
    "                %90th percentile\n",
    "                metrics.t2R_stats(k,4) = prctile( tmp_data,90 );\n",
    "\n",
    "                %max\n",
    "                metrics.t2R_stats(k,5) = max( tmp_data );\n",
    "            end\n",
    "\n",
    "        end\n",
    "        \n",
    "    %Compute the min, 10th, 50th, 90th percentile, and max values of\n",
    "    %resistance for each year/cohort. If resistance values are all zeros,\n",
    "    %this cohort/year is ignored (and the value is left blank but not 0).\n",
    "\tmetrics.yearly_freq(:,1) = num2cell( data_store(:,1)+1 );\n",
    "        for y = 1:nCohorts*nYears\n",
    "            %\n",
    "            ind = find(data_store(y,2:end)~=0);\n",
    "            if ~isempty(ind)\n",
    "                metrics.yearly_freq{y,2} = min(data_store(y,ind+1),[],2);\n",
    "                metrics.yearly_freq{y,3} = prctile(data_store(y,ind+1),10,2);\n",
    "                metrics.yearly_freq{y,4} = prctile(data_store(y,ind+1),50,2);\n",
    "                metrics.yearly_freq{y,5} = prctile(data_store(y,ind+1),90,2);\n",
    "                metrics.yearly_freq{y,6} = max(data_store(y,ind+1),[],2);\n",
    "            end\n",
    "        end\n",
    "    \n",
    "    %% Write data to Excel file\n",
    "    \n",
    "    %Get reference to open Excel sheet (still open from parameter entry) \n",
    "    h = actxGetRunningServer('Excel.Application');\n",
    "    myBook = h.Workbooks.Item('WeedResistanceGUI.xlsm');\n",
    "    \n",
    "    %Write metrics to \"Results\" tab\n",
    "    mySheet = myBook.Sheets.Item('Results');\n",
    "    \n",
    "    %Clear old data (Excel is already doing this, so it is commented out\n",
    "    %here)\n",
    "%     mySheet.Range('F7:F13').Value = '';\n",
    "%     mySheet.Range('C17:D1017').Value = '';\n",
    "%     mySheet.Range('H8:M1017').Value = '';\n",
    "%     \n",
    "%     %Number of simulations\n",
    "%     mySheet.Range('F7').Value = nSim;\n",
    "    \n",
    "    %Probability of exceeding resistances thresholds\n",
    "    mySheet.Range('F8').Value = metrics.probs(1);\n",
    "    mySheet.Range('F9').Value = metrics.probs(2);\n",
    "    mySheet.Range('F10').Value = metrics.probs(3);\n",
    "    \n",
    "    %Median times to threshold (convert to string in case value = Inf)\n",
    "    mySheet.Range('F11').Value = num2str( metrics.t2R_stats(1,3) );\n",
    "    mySheet.Range('F12').Value = num2str( metrics.t2R_stats(2,3) );\n",
    "    mySheet.Range('F13').Value = num2str( metrics.t2R_stats(3,3) );\n",
    "    \n",
    "    %Average integrated risk\n",
    "    mySheet.Range('F14').Value = mean(metrics.IRF);\n",
    "    \n",
    "    %Individual simulation integrated risk\n",
    "    n = 1:nSim;\n",
    "    mySheet.Range( strcat('I18:I',num2str(17+nSim)) ).Value = n';\n",
    "    mySheet.Range( strcat('J18:J',num2str(17+nSim)) ).Value = metrics.IRF;\n",
    "    \n",
    "    %Resistance versus time statistics\n",
    "    [rows,~] = size(metrics.yearly_freq);\n",
    "    mySheet.Range( strcat('B18:B',num2str(17+rows)) ).Value = metrics.yearly_freq(:,1);\n",
    "    mySheet.Range( strcat('C18:G',num2str(17+rows)) ).Value = metrics.yearly_freq(:,2:end);\n",
    "    \n",
    "    %Write raw data (upper and lower seed bank resistant fraction) to text\n",
    "    %file\n",
    "    dlmwrite('UpperBankFreq.csv',freq_output.upperbank_freq,',');\n",
    "    dlmwrite('LowerBankFreq.csv',freq_output.lowerbank_freq,',');\n",
    "    dlmwrite('UpperBankDens.csv',dens_output.upperbank_freq,',');\n",
    "    dlmwrite('LowerBankDens.csv',dens_output.lowerbank_freq,',');\n",
    "    \n",
    "end\n",
    "\n",
    "function [tR,counter] = getResistanceTime(time,res_frac,f,varargin)\n",
    "    %time is a (Mx1) array. res_frac is an (Mx1) array. time and res_frac\n",
    "    %must have the same number of rows, otherwise an error is returned. f\n",
    "    %is the resistant fraction whose time-to-resistance will be found. f\n",
    "    %may be a (Px1) array. Function returns the time-to-resistance, tR,\n",
    "    %which is a (Px1) array. varargin is a Px1 array which may be used to\n",
    "    %keep \"count\" of how many times resistance developed for each element\n",
    "    %of f. If resistance is found for current value of f, a counter\n",
    "    %variable is incremented by 1. This is then returned in counter. This\n",
    "    %is useful for counting how many times resistance developed over many\n",
    "    %function calls\n",
    "    \n",
    "    if length(time) ~= length(res_frac)\n",
    "        error('time and res_frac must have equal number of rows');\n",
    "        \n",
    "    else\n",
    "        %Counter variable\n",
    "        if ~isempty(varargin)\n",
    "            counter = varargin{1};\n",
    "        else\n",
    "            counter = zeros(length(f),1);\n",
    "        end\n",
    "        \n",
    "        %number of evaluation points, P\n",
    "        P = length(f);\n",
    "        \n",
    "        %output variable, a (Px1) array\n",
    "        tR = zeros(P,1);\n",
    "        \n",
    "        %Loop through f and find corresponding times.\n",
    "        for j = 1:P\n",
    "            tmp = find(res_frac>f(j),1,'first');\n",
    "\n",
    "            if tmp == 1\n",
    "                %The initial resistance exceeded the threshold\n",
    "                tR(j) = 0;\n",
    "                \n",
    "                %Increment counter\n",
    "                counter(j) = counter(j) + 1;\n",
    "            elseif ~isempty(tmp)\n",
    "                %If the threshold is exceeded, interpolate to find the time\n",
    "\n",
    "                %Temporary time and resistant fraction variables for\n",
    "                %readability\n",
    "                t2 = time(tmp);\n",
    "                t1 = time(tmp-1);\n",
    "                f2 = res_frac(tmp);\n",
    "                f1 = res_frac(tmp-1);\n",
    "\n",
    "                %Linear interpolation\n",
    "                tR(j) = (t2 - t1) / (f2 - f1) * (f(j) - f1) + t1;\n",
    "                \n",
    "                %Increment counter\n",
    "                counter(j) = counter(j) + 1;\n",
    "            else\n",
    "                %The threshold was never exceeded\n",
    "                tR(j) = Inf;\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function [ ] = samplePlots( )\n",
    "%% Initialization\n",
    "\n",
    "P = readData();\n",
    "\n",
    "%create figure\n",
    "figure('Position',get(0,'ScreenSize'))\n",
    "\n",
    "%choose lifestage to plot\n",
    "lifestage = 'seedBank';\n",
    "\n",
    "h = waitbar(0, 'Please wait', 'Name', 'Plot progress', ...\n",
    "    'CreateCancelBtn', 'setappdata(gcbf, ''canceling'', 1)');\n",
    "setappdata(h, 'canceling', 0)\n",
    "\n",
    "%run simulation number of times specified and plot emergence curve for each\n",
    "subplot(2,2,1)\n",
    "hold all\n",
    "nSim = P.General.nSim;      %number of simulations to run\n",
    "nRes = 0;       %number of times resistance develops after nYears\n",
    "\n",
    "\n",
    "%% Run simulation specified number of times \n",
    "\n",
    "%Resistant fraction evaulation points\n",
    "fEval = 0.05:0.05:0.95;\n",
    "yearEval = zeros(nSim,length(fEval));\n",
    "\n",
    "%Default threshold for when population is considered resistant\n",
    "toler = P.General.criticalResTime;\n",
    "\n",
    "%Create array to store resistance metric data for all simulations\n",
    "resData = zeros(nSim, 3);\n",
    "\n",
    "%Run simulation\n",
    "for n = 1:nSim\n",
    "\n",
    "    if getappdata(h, 'canceling')\n",
    "        break\n",
    "    end\n",
    "    \n",
    "    [Population,Params] = Main_2();\n",
    "    plotPopulation(Population.(lifestage),Params.key,'fracR');\n",
    "    \n",
    "    %Compute resistance risk: fraction of times resistant population is >= 25% of\n",
    "    %total after N years, (defined by nYears in defineParams() function)\n",
    "    R = sumResistant(Population.(lifestage)(:,:,end),Params.key); %resistant population\n",
    "    T = sum(Population.(lifestage)(:,:,end),2);    %total population (at end of year)\n",
    "    \n",
    "    if R(end)/T(end) >= 0.25\n",
    "        nRes = nRes + 1;\n",
    "    end\n",
    "\n",
    "    %Create variable for storing the resistance risk using the Population\n",
    "    %parameter (first time through loop only)\n",
    "    if n==1\n",
    "        ResistanceRisk = zeros(1,Params.General.nYears);\n",
    "    end\n",
    "\n",
    "    %count the resistants in each year and divide by total population\n",
    "    resPop = sumResistant(Population.(lifestage)(:,:,end),Params.key);\n",
    "    totPop = sum(Population.(lifestage)(:,:,end),2);\n",
    "    resFrac = resPop./totPop;\n",
    "\n",
    "    %find the years when the resistant fraction exceeds the tolerance\n",
    "    ind = find(resFrac>=toler);\n",
    "    ResistanceRisk(ind) = ResistanceRisk(ind) + 1;\n",
    "\n",
    "    %find fractional year values corresponding to the\n",
    "    %resistant_fraction array specified by the user\n",
    "    yearEval(n,:) = sigmoidMetrics(0:length(resFrac)-1,resFrac,fEval);\n",
    "\n",
    "    %Store resistance metric data for this simulation\n",
    "    resData(n, :) = Population.Res;\n",
    "    \n",
    "    %Update the progress bar\n",
    "    waitbar(n / nSim)\n",
    "end\n",
    "\n",
    "delete(h)\n",
    "\n",
    "%% Store resistance metric data in Excel\n",
    "\n",
    "range = strcat('D17:F', num2str(17+nSim-1));\n",
    "h = actxGetRunningServer('Excel.Application');\n",
    "myBook = h.Workbooks.Item('WeedResistance_ParameterEntryInterface.xlsm');\n",
    "mySheet = myBook.Sheets.Item('Results');\n",
    "mySheet.Range(range).Value = resData;\n",
    "\n",
    "%% Add title and resistance risk text to first plot\n",
    "title(strcat('Fraction resistant: ',num2str(nSim), ' simulations'))\n",
    "pos = Params.General.nYears / 2;\n",
    "text(pos,0.2,strcat('Resistance risk = ',num2str(nRes/nSim)));\n",
    "\n",
    "%% Plots with +- one standard deviation window\n",
    "\n",
    "%convert ResistanceRisk to a probability\n",
    "ResistanceRisk = ResistanceRisk/nSim;\n",
    "\n",
    "%eliminate curves where didn't or partially developed\n",
    "k = 1;\n",
    "while k <= size(yearEval,1)\n",
    "    if sum(isnan(yearEval(k,:)))~=0\n",
    "        yearEval(k,:) = [];\n",
    "    else\n",
    "        k = k + 1;\n",
    "    end\n",
    "end\n",
    "\n",
    "%Take average, standard deviation, and store resistant_fraction in array\n",
    "%meanCurve, for output\n",
    "meanCurve(:,1) = mean(yearEval, 1);\n",
    "meanCurve(:,2) = std(yearEval, 0, 1);\n",
    "meanCurve(:,3) = fEval;\n",
    "\n",
    "%Fit a sigmoid to the output (using subfunction sigmoidFun, below)\n",
    "p = lsqcurvefit(@sigmoidFun,[0.5,5],meanCurve(:,1),meanCurve(:,3));\n",
    "\n",
    "%Use fit parameters to create a new population variable pop, which may\n",
    "%be smoother than the original simulation data.\n",
    "t = linspace(0,Params.General.nYears-1);\n",
    "pop = sigmoidFun(p,t);\n",
    "t = t';\n",
    "pop = pop';\n",
    "\n",
    "% ax1 = axes('Position',[0.1,0.1,0.38,0.85]);\n",
    "subplot(2,2,2)\n",
    "hold all\n",
    "% set(0,'DefaultAxesColorOrder',[0 0 1;1 0 0;0 0.4 0;1 0.6 0])\n",
    "\n",
    "%lower window\n",
    "p = lsqcurvefit(@sigmoidFun,[0.5,5],meanCurve(:,1)-2*meanCurve(:,2),meanCurve(:,3));\n",
    "wind = sigmoidFun(p,t);\n",
    "area(t,wind,'FaceColor',[0.7,0.7,0.7],'LineStyle','none')\n",
    "plot(t,wind,'k','LineWidth',1,'LineStyle',':')\n",
    "\n",
    "%upper window\n",
    "p = lsqcurvefit(@sigmoidFun,[0.5,5],meanCurve(:,1)+2*meanCurve(:,2),meanCurve(:,3));\n",
    "wind = sigmoidFun(p,t);\n",
    "area(t,wind,'FaceColor',[1,1,1],'LineStyle','none')\n",
    "plot(t,wind,'k','LineWidth',1,'LineStyle',':')\n",
    "\n",
    "plot(t,pop,'k-','LineWidth',3)\n",
    "%plot(t,-2*wind,'LineWidth',3,'LineStyle',':')\n",
    "xlabel('Year','Interpreter','latex')\n",
    "ylabel('Fraction resistant','Interpreter','latex')\n",
    "ax = gca;\n",
    "ax.FontSize = 16;\n",
    "\n",
    "%% Resistance risk versus time plot\n",
    "\n",
    "%Make data look smooth using Piecewise Cubic Hermite Interpolating\n",
    "%Polynomial (PCHIP)\n",
    "\n",
    "%Years in original simulation\n",
    "n = 0:length(ResistanceRisk)-1;\n",
    "\n",
    "%Points to do the PCHIP through\n",
    "t = linspace(0,max(n));\n",
    "\n",
    "%Call pchip\n",
    "pop = pchip(n,ResistanceRisk,t);\n",
    "\n",
    "%Plot results\n",
    "t = t';\n",
    "pop = pop';\n",
    "\n",
    "% ax2 = axes('Position',[0.6,0.1,0.38,0.85]);\n",
    "subplot(2,2,3)\n",
    "hold all\n",
    "plot(t,pop*100,'k-','LineWidth',3,'MarkerSize',20)\n",
    "xlabel('Year','Interpreter','latex')\n",
    "ylabel('Resistance risk $\\%$','Interpreter','latex')\n",
    "ax = gca;\n",
    "ax.FontSize = 16;\n",
    "\n",
    "end\n",
    "\n",
    "function y = sigmoidFun(x,xdata)\n",
    "    y = 1 ./ (1 + exp(-x(1).*xdata + x(2)));\n",
    "end\n",
    "\n",
    "function Rpop = sumResistant(pop,key)\n",
    "\n",
    "    %Determine number of gene loci and genotypes from key variable\n",
    "    [nLoci,nGeno] = size(key);\n",
    "    \n",
    "    %Determine number of years and cohorts from population variable\n",
    "    [nYears,~,nCohorts] = size(pop);\n",
    "    \n",
    "    %Initialize output array to 0\n",
    "    Rpop = zeros(nYears,nCohorts);\n",
    "    \n",
    "    %Sum all with resistant allele\n",
    "    for G = 1:nGeno\n",
    "        for L = 1:nLoci\n",
    "            if key(L,G) == 2 || key(L,G) == 1\n",
    "                Rpop(:,:) = Rpop(:,:) + squeeze(pop(:,G,:));\n",
    "                break\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "    \n",
    "end\n",
    "\n",
    "function nEval = sigmoidMetrics(n,f,fEval)\n",
    "    \n",
    "    %Number of evaluation points\n",
    "    len = length(fEval);\n",
    "\n",
    "    %Array to store output\n",
    "    nEval = zeros(1,len);\n",
    "    \n",
    "    %Loop through evaluation points and populate nEval\n",
    "    for k = 1:len\n",
    "        \n",
    "        %Find the first value >= current evaluation point\n",
    "        ind = find(f>=fEval(k),1,'first');\n",
    "        \n",
    "        %If no value found or if the population initially exceeds\n",
    "        %tolerance, store NaN, otherwise, find the time corresponding to\n",
    "        %the current evaulation point (using linear interpolation)\n",
    "        if isempty(ind)\n",
    "            nEval(k) = NaN;\n",
    "        elseif ind == 1\n",
    "            nEval(k) = NaN;\n",
    "        else\n",
    "            nEval(k) = (n(ind)-n(ind-1))/(f(ind)-f(ind-1))*(fEval(k)-f(ind-1)) + n(ind-1);\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function seedIn = seedImmigration(nGeno,varargin)\n",
    "    seedIn = zeros(1,nGeno);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function newPop = selectiveSurvival(oldPop,fSurvival,targetLocus,key)\n",
    "%fSurvival = 3 x 1 array containing: [fSS,fRS,fRR]\n",
    "[nLoci,nGeno] = size(key);\n",
    "newPop = oldPop*0;\n",
    "\n",
    "if targetLocus > nLoci\n",
    "    warning('Target locus exceeds the number specified by the user in the input file, original population returned');\n",
    "    newPop = oldPop;\n",
    "else\n",
    "   for k = 1:nGeno\n",
    "       switch key(targetLocus,k)\n",
    "           case 0\n",
    "               %homozygous susceptible\n",
    "                newPop(k) = fSurvival(1)*oldPop(k);\n",
    "           case 1\n",
    "               %heterozygous\n",
    "               newPop(k) = fSurvival(2)*oldPop(k);\n",
    "           case 2\n",
    "               newPop(k) = fSurvival(3)*oldPop(k);\n",
    "       end\n",
    "       \n",
    "   end\n",
    "   \n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function nEval = sigmoidMetrics(n,f,fEval)\n",
    "    \n",
    "    %Number of evaluation points\n",
    "    len = length(fEval);\n",
    "\n",
    "    %Array to store output\n",
    "    nEval = zeros(1,len);\n",
    "    \n",
    "    %Loop through evaluation points and populate nEval\n",
    "    for k = 1:len\n",
    "        \n",
    "        %Find the first value >= current evaluation point\n",
    "        ind = find(f>=fEval(k),1,'first');\n",
    "        \n",
    "        %If no value found or if the population initially exceeds\n",
    "        %tolerance, store NaN, otherwise, find the time corresponding to\n",
    "        %the current evaulation point (using linear interpolation)\n",
    "        if isempty(ind)\n",
    "            nEval(k) = NaN;\n",
    "        elseif ind == 1\n",
    "            nEval(k) = NaN;\n",
    "        else\n",
    "            nEval(k) = (n(ind)-n(ind-1))/(f(ind)-f(ind-1))*(fEval(k)-f(ind-1)) + n(ind-1);\n",
    "        end\n",
    "        \n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function Rpop = sumResistant(pop,key,varargin)\n",
    "\n",
    "if isempty(varargin)\n",
    "    sumOver = 'all';\n",
    "else\n",
    "    sumOver = varargin{1};\n",
    "end\n",
    "\n",
    "%determine number of gene loci and genotypes from key variable\n",
    "[nLoci,nGeno] = size(key);\n",
    "\n",
    "%determine number of years and cohorts from population variable\n",
    "[nYears,~,nCohorts] = size(pop);\n",
    "\n",
    "%initialize output array to 0\n",
    "Rpop = zeros(nYears,nCohorts);\n",
    "\n",
    "switch sumOver\n",
    "\n",
    "    case 'homozygous'\n",
    "        for G = 1:nGeno\n",
    "            for L = 1:nLoci\n",
    "                if key(L,G) == 2 \n",
    "                    Rpop(:,:) = Rpop(:,:) + squeeze(pop(:,G,:));\n",
    "                    break\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "    otherwise\n",
    "        for G = 1:nGeno\n",
    "            for L = 1:nLoci\n",
    "                if key(L,G) == 2 || key(L,G) == 1\n",
    "                    Rpop(:,:) = Rpop(:,:) + squeeze(pop(:,G,:));\n",
    "                    break\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "        \n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "function newPopulation = survival(oldPopulation,fSurvival)\n",
    "    newPopulation = fSurvival*oldPopulation;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating a variable and storing the text\n",
    "# that we want to search\n",
    "def search_replace(filename,search, replace):\n",
    "\tsearch_text = search\n",
    "\n",
    "\t# creating a variable and storing the text\n",
    "\t# that we want to add\n",
    "\treplace_text = replace\n",
    "\n",
    "\t# Opening our text file in read only\n",
    "\t# mode using the open() function\n",
    "\twith open(filename, 'r') as file:\n",
    "\n",
    "\t\t# Reading the content of the file\n",
    "\t\t# using the read() function and storing\n",
    "\t\t# them in a new variable\n",
    "\t\tdata = file.read()\n",
    "\n",
    "\t\t# Searching and replacing the text\n",
    "\t\t# using the replace() function\n",
    "\t\tdata = data.replace(search_text, replace_text)\n",
    "\n",
    "\t# Opening our text file in write only\n",
    "\t# mode to write the replaced content\n",
    "\tfilename = filename.split('.')[0]+'.py'\n",
    "\twith open(filename, 'w') as file:\n",
    "\n",
    "\t\t# Writing the replaced data in our\n",
    "\t\t# text file\n",
    "\t\tfile.write(data)\n",
    "\n",
    "\t# Printing Text replaced\n",
    "\tprint(\"Text replaced\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "competition.py\n",
      "Text replaced\n",
      "deepTill.py\n",
      "Text replaced\n",
      "extinction.py\n",
      "Text replaced\n",
      "getResistanceTime.py\n",
      "Text replaced\n",
      "Herbicide_Data.py\n",
      "Text replaced\n",
      "LowerBankDens.py\n",
      "Text replaced\n",
      "LowerBankFreq.py\n",
      "Text replaced\n",
      "Main.py\n",
      "Text replaced\n",
      "MainSimulation.py\n",
      "Text replaced\n",
      "mating.py\n",
      "Text replaced\n",
      "mutation.py\n",
      "Text replaced\n",
      "PackagingLog.py\n",
      "Text replaced\n",
      "plotPopulation.py\n",
      "Text replaced\n",
      "prepareParams_2.py\n",
      "Text replaced\n",
      "readData.py\n",
      "Text replaced\n",
      "resistanceRisk.py\n",
      "Text replaced\n",
      "resMetrics.py\n",
      "Text replaced\n",
      "runSimulation.py\n",
      "Text replaced\n",
      "runSimulation_old_working.py\n",
      "Text replaced\n",
      "samplePlots.py\n",
      "Text replaced\n",
      "seedImmigration.py\n",
      "Text replaced\n",
      "selectiveSurvival.py\n",
      "Text replaced\n",
      "selfTable1.py\n",
      "Text replaced\n",
      "selfTable2.py\n",
      "Text replaced\n",
      "selfTable3.py\n",
      "Text replaced\n",
      "selfTable4.py\n",
      "Text replaced\n",
      "selfTable5.py\n",
      "Text replaced\n",
      "sigmoidMetrics.py\n",
      "Text replaced\n",
      "sumResistant.py\n",
      "Text replaced\n",
      "survival.py\n",
      "Text replaced\n",
      "test.py\n",
      "Text replaced\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "for i in os.listdir('BMP_Matlab_Code/'):\n",
    "    if i.endswith('.py'):\n",
    "        print(i)\n",
    "        search_replace('BMP_Matlab_Code/'+i,'zeros','np.zeros')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "d0772284b9d471da871167e9b1d70e67ab60113943bf15725750ad854e805c67"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
